sdk/aqueduct/models/dag.py

```diff
@@ -47,21 +50,113 @@ class DAG(BaseModel):
+        the operators in topological order so that numbers are assigned in a reasonable fashion. Output artifacts
+        of the same operator are always numbered consecutively.
+
+        `seen_artifact_names` should have already been updated by `_validate_explicitly_named_artifacts`.
+        """
+        q: List[Operator] = copy.copy(starting_ops)
+        seen_op_ids: List[uuid.UUID] = []
+
+        any_op_was_renamed = False
+        while len(q) > 0:
+            curr_op = q.pop(0)
+
+            # Only traverse operators you haven't seen before.
+            if curr_op.id in seen_op_ids:
+                continue
+            seen_op_ids.append(curr_op.id)
+
+            output_artifacts = self.must_get_artifacts(curr_op.outputs)
+            for artifact in output_artifacts:
+                # Skip name resolution for explicitly named artifacts.
+                # We've already checked in the first pass.
+                if not artifact.explicitly_named:
+                    # Find an unallocated name for each artifact.
+                    original_name = artifact.name
+                    while artifact.name in seen_artifact_names:
+                        artifact.name = bump_artifact_suffix(artifact.name)
+
+                    if original_name != artifact.name:
+                        logger().warning(
+                            "Multiple artifacts were named `%s`. Since artifact names must be unique, "
+                            "we renamed one of them to `%s`." % (original_name, artifact.name)
+                        )
+                        any_op_was_renamed = True
+
+                    seen_artifact_names.add(artifact.name)
+
+                q.extend(self.list_operators(on_artifact_id=artifact.id))
+
+        if any_op_was_renamed:
+            logger().warning(
+                "Note that any artifacts you explicitly gave a name to were not renamed."
+            )
+
+    def validate_and_resolve_artifact_names(self) -> None:
+        """To be called from publish_flow() only.
+
+        Checks that all explicitly named artifacts are unique.
+
+        For implicitly named artifacts that collide, bumps their names using the `(num)` suffix.
+        The index is grouped by the input operator, so an operator with multiple output artifacts
+        that use our default naming scheme will always have consecutive numbers. Eg. (1), (2), (3)
+        """
+
+        # For some determinism around naming, we sort the starting operators by name and then id.
+        starting_ops: List[Operator] = sorted(
+            [op for op in self.operators.values() if len(op.inputs) == 0],
+            key=lambda op: (op.name, op.id),
+        )
+
+        seen_artifact_names: Set[str] = set()
+        self._validate_explicitly_named_artifacts(starting_ops, seen_artifact_names)
+        self._resolve_artifact_names(starting_ops, seen_artifact_names)
```

Comment:
The main reason is that we want the first pass to only handle explicitly named artifacts so that those artifacts can claim their names first. Now that I think about it, only the second pass needs to be topologically ordered, since that's the one that's bumping the suffix. Can fix.